### DISCUSSION TITLE
Adonis 5, implement custom auth provider / guard

### DISCUSSION THREAD

**Discussion by skrenek** [adonisjs/core#1490] [Help]
I'm looking for some insights as to the best way to handle something in Adonis 5.  I've implemented an auth scheme and serializer in Adonis 4.1 to support verifying Auth0 JWT tokens, and need to do something similar in an Adonis 5 experiment.  

The app I'm experimenting with has the flow of a Vue.js app which authenticates with Auth0 on the UI side.  I'm writing a possible replacement API that will accept the JWT token Auth0 returns and verify it with Auth0 to prevent the user from having to login twice.  

I've read https://github.com/adonisjs/core/discussions/1365, but that discussion seems to be finished so I'm starting this as a separate conversation. 

It appears I would need to create a custom Auth guard in Adonis 5 to do the same thing I've done in Adonis 4, but it is not obvious to me the best way to go about this.  Is there a good way to hook into as much of the Adonis auth system as I can, or am I better off doing this outside the Adonis 5 auth flow and just writing my own middleware?

**Comment by alex-oliveira** [comment]
I'm also looking for a way to use JWT integrated in Adonis. But for me it's not obvious how I should do this either. Have you resolved this issue?

**Comment by sasha172017** [comment]
I'm looking for a way to use JWT  too

**Comment by skrenek** [comment]
I was able to find a solution, although I'm not sure if it's the best way.  I ended up creating a custom authentication middleware that verifies the token.  This may or may not work for you.  Because we're working with Auth0, it's really only a partial solution as we don't have to worry about generating the tokens themselves.

The middleware I wrote does the following:

- reads Auth0's access token off of the request's Authorization header
- Uses the `jsonwebtoken` module's verify function to verify the token.  I store the private cert from Auth0 used to verify it in an env variable and inject it through the standard Adonis config mechanisms
- If any of those steps fail, the middleware responds with response.unauthorized.

I recognize thetutlage's concerns with JWT in general vs opaque tokens, but in my case the system uses Auth0 and as such I really have no other option but to use them.  We're not using any data in the tokens, but are using them purely for authentication so I feel any risk is minimal.

**Reply by sasha172017** [reply]
this is not best way. Best way do it use adonis/auth with custom driver jwt. We need docs by custom driver, guard. need to wait for developers the framework. I hope to wait a little

**Reply by skrenek** [reply]
I agree.  It is not the best way, but as of last August it was the only way I knew.  As documentation improves and I get time I will be replacing it, but until then it's working. :)

**Comment by ad-on-is** [comment]
Maybe this helps. I'm working on a project for a while now with v5 and have implemented token-side login with facebook/google/apple on my own.

My workflow is as follows.

- Login user via Social auth on the client and get the `access_token`
- Send ONLY the access token to AdonisJs server and validate it via the social API's validation URL
- If the token is valid, the social API returns the e-mail address
- If the user does not exist in the DB, then I save the user with email = social.email and password = hash(social.email)
- Finally I return the AdonisJs token

```js
  const tokenData = await auth.use('api').attempt(authData.email, authData.password)
  return response.safeHeader('Bearer-Token', token.token).accepted(null)
```

I know, setting the password the same as the email might be a stupid idea, but my app does not offer login with username/password, only with social media.

**Comment by alex-oliveira** [comment]
I did an implementation using adonis guard, which I consider to be the "most correct" way.

https://github.com/adonisjs/core/discussions/2506#discussioncomment-825904

**Reply by skrenek** [reply]
Thanks for sharing, Alex.  I'll take a look in the near future.
